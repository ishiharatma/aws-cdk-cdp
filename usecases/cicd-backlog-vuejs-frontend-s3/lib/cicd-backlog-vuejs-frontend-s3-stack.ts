import * as cdk from 'aws-cdk-lib';
import { StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';
import {
  aws_s3 as s3,
  aws_lambda as lambda,
  aws_apigateway as apigw,
  aws_iam as iam,
  aws_codedeploy as codedeploy,
  aws_codepipeline as codepipeline,
  aws_codepipeline_actions as codepipeline_actions,
  aws_codebuild as codebuild,
 } from 'aws-cdk-lib';

interface CicdBacklogVuejsFrontendS3StackProps extends StackProps {
  readonly pjName: string;
  readonly envName: string;
  readonly repositoryName: string;
  readonly branchName: string;
  readonly allowedIps?: string[];
  readonly websiteBucketName?: string;
  readonly distributionId?: string;
  readonly isAutoDeleteObject?: boolean;
}


export class CicdBacklogVuejsFrontendS3Stack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CicdBacklogVuejsFrontendS3StackProps) {
    super(scope, id, props);

    const accountId = cdk.Stack.of(this).account;

    // Create Source S3 bucket
    const bucketName = `${props.pjName}-${props.envName}-source-s3-${accountId}`;
    const bucket = new s3.Bucket(this, bucketName, {
      bucketName: bucketName,
      removalPolicy: props.isAutoDeleteObject ? cdk.RemovalPolicy.DESTROY : cdk.RemovalPolicy.RETAIN,
      autoDeleteObjects: props.isAutoDeleteObject,
    });

    // Lambda function name
    const gitCloneFunctionName = `${props.pjName}-${props.envName}-git-clone-lambda-${accountId}`;

    // create lambda Iam Role and policy
    const lambdaRole = new iam.Role(this, `${gitCloneFunctionName}-lambda-role`, {
      roleName: `${gitCloneFunctionName}-role`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });
    lambdaRole.attachInlinePolicy(new iam.Policy(this, `${gitCloneFunctionName}-lambda-policy`, {
      policyName: `${gitCloneFunctionName}-policy`,
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
          ],
          resources: [
            `arn:aws:logs:${cdk.Stack.of(this).region}:${accountId}:log-group:/aws/lambda/${props.pjName}-${props.envName}-*`,
          ],
        }),
      ],
    }));
    // attach policy to lambda role for s3
    lambdaRole.attachInlinePolicy(new iam.Policy(this, `${gitCloneFunctionName}-s3-policy`, {
      policyName: `${gitCloneFunctionName}-s3-policy`,
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            's3:GetObject',
            's3:PutObject',
            's3:ListBucket',
          ],
          resources: [
            bucket.bucketArn,
            `${bucket.bucketArn}/*`,
          ],
        }),
      ],
    }));  

    // Create Lambda function for git clone
    const pipelineName = `${props.pjName}-${props.envName}-pipeline`;
    const gitCloneFunctionPath = path.join(__dirname, '../src/lambda/backlog-git-clone/python');
    const gitCloneFunction = new lambda.Function(this, gitCloneFunctionName, {
      functionName: gitCloneFunctionName,
      description: 'git clone lambda function for s3',
      runtime: lambda.Runtime.PYTHON_3_13,
      handler: 'index.lambda_handler',
      timeout: cdk.Duration.seconds(300),
      memorySize: 1024,
      role: lambdaRole,
      architecture: lambda.Architecture.ARM_64,
      code: lambda.Code.fromAsset(gitCloneFunctionPath),
      layers: [
        lambda.LayerVersion.fromLayerVersionArn(this, 'LambdaLayer', `arn:aws:lambda:ap-northeast-1:${accountId}:layer:dulwich-layer:2`),
      ],
      environment: {
        BUCKET_NAME: bucket.bucketName,
        REPOSITORY: props.repositoryName,
        BRANCH: props.branchName,
        ZIP_FILE_NAME: "source",
        USER: "",
        PASS: "",
        PIPELINE_NAME: pipelineName,
      },
    });

    // Create API Gateway for Lambda
    const api = new apigw.LambdaRestApi(this, 'GitCloneApi', {
      restApiName: 'GitCloneApi',
      description: 'This service is for git clone',
      handler: gitCloneFunction,
      deployOptions: {
        stageName: 'dev',
      },
      policy: props.allowedIps ? new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            effect: iam.Effect.DENY,
            principals: [new iam.AnyPrincipal()],
            actions: ['execute-api:Invoke'],
            resources: ['execute-api:/*/*/*'], // ステージ作成前なのでワイルドカードを使用
            conditions: {
              'NotIpAddress': {
                'aws:SourceIp': props.allowedIps,
              },
            },
          }),
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            principals: [new iam.AnyPrincipal()],
            actions: ['execute-api:Invoke'],
            resources: ['execute-api:/*/*/*'],
          }),
        ],
      }) : undefined,
    });
    // add resource /backlog-webhook
    // https://[API_ID].execute-api.[REGION].amazonaws.com/dev/backlog-git-webhook
    const webhook = api.root.addResource('backlog-git-webhook');
    // add method POST
    webhook.addMethod('POST');

    // output api url
    new cdk.CfnOutput(this, 'ApiUrl', { value: api.url });

    // 静的ウェブサイトのバケット
    let websiteBucket: s3.IBucket;
    if (props.websiteBucketName) {
      // 既存のウェブサイトバケットを参照
      websiteBucket = s3.Bucket.fromBucketName(this, 'ExistingWebsiteBucket', props.websiteBucketName);
    } else {
      // 新しいウェブサイトバケットを作成
      const websiteBucketName = `${props.pjName}-${props.envName}-website-${accountId}`;
      websiteBucket = new s3.Bucket(this, websiteBucketName, {
        bucketName: websiteBucketName,
        websiteIndexDocument: 'index.html',
        websiteErrorDocument: 'index.html',
        publicReadAccess: true,
        removalPolicy: props.isAutoDeleteObject ? cdk.RemovalPolicy.DESTROY : cdk.RemovalPolicy.RETAIN,
        autoDeleteObjects: props.isAutoDeleteObject,
        cors: [
          {
            allowedMethods: [
              s3.HttpMethods.GET,
              s3.HttpMethods.HEAD,
            ],
            allowedOrigins: ['*'],
            allowedHeaders: ['*'],
          },
        ],
      });
    }

    // パイプラインの変数定義を作成
    const repositoryNameVar = new codepipeline.Variable({
      variableName: 'REPOSITORY_NAME',
      description: 'The name of the repository',
      defaultValue: props.repositoryName || '',
    });
    const branchNameVar = new codepipeline.Variable({
      variableName: 'BRANCH_NAME',
      description: 'The branch name of the repository',
      defaultValue: props.branchName || '',
    });
    const commitIdVar = new codepipeline.Variable({
      variableName: 'COMMIT_ID',
      description: 'The commit ID of the repository',
      defaultValue:'',
    });

    // Create CodePipeline
    const pipeline = new codepipeline.Pipeline(this, 'Pipeline', {
      pipelineName: pipelineName,
      pipelineType: codepipeline.PipelineType.V2,
      restartExecutionOnUpdate: true,
      variables: [repositoryNameVar, branchNameVar, commitIdVar],
    });

    // Source stage
    const sourceOutput = new codepipeline.Artifact('SourceOutput');
    const sourceAction = new codepipeline_actions.S3SourceAction({
      actionName: 'S3Source',
      bucket: bucket,
      bucketKey: 'source.zip',
      output: sourceOutput,
      trigger: codepipeline_actions.S3Trigger.NONE, // Webhook Lambda から手動でトリガー
    });
    // Add stages to pipeline
    pipeline.addStage({
      stageName: 'Source',
      actions: [sourceAction],
    });

    // Create CodeBuild project
    const buildProject = new codebuild.PipelineProject(this, 'BuildProject', {
      projectName: `${props.pjName}-${props.envName}-build`,
      environment: {
        buildImage: codebuild.LinuxBuildImage.AMAZON_LINUX_2_4,
        privileged: true,
      },
      buildSpec: codebuild.BuildSpec.fromSourceFilename('buildspec.yml'),
      cache: codebuild.Cache.local(codebuild.LocalCacheMode.CUSTOM),
//      buildSpec: codebuild.BuildSpec.fromObject({
//        version: '0.2',
//        phases: {
//          install: {
//            'runtime-versions': {
//              nodejs: '18',
//            },
//            commands: [
//              'echo Installing dependencies...',
//              'npm install',
//            ],
//          },
//          build: {
//            commands: [
//              'echo Building Vue.js application...',
//              'npm run build',
//            ],
//          },
//          post_build: {
//            commands: [
//              'echo Build completed successfully',
//            ],
//          },
//        },
//        artifacts: {
//          'base-directory': 'dist',
//          files: [
//            '**/*',
//          ],
//        },
//      }),
    });

    // Build stage
    const buildOutput = new codepipeline.Artifact('BuildOutput');
    const buildAction = new codepipeline_actions.CodeBuildAction({
      actionName: 'Build',
      project: buildProject,
      input: sourceOutput,
      outputs: [buildOutput],
    });
    pipeline.addStage({
      stageName: 'Build',
      actions: [buildAction],
    });

    // Deploy stage
    const deployAction = new codepipeline_actions.S3DeployAction({
      actionName: 'PublishAssets',
      bucket: websiteBucket,
      input: buildOutput,
      accessControl: s3.BucketAccessControl.PUBLIC_READ,
      extract: true,
    });

    pipeline.addStage({
      stageName: 'Deploy',
      actions: [deployAction],
    });

    // Add permissions for pipeline to access S3 bucket
    bucket.grantRead(pipeline.role);
    websiteBucket.grantWrite(pipeline.role);

    // Cleanup S3 buclet
    const cleanupFunctionName = `${props.pjName}-${props.envName}-cleanup-lambda-${accountId}`;
    const lambdaRoleCleanup = new iam.Role(this, `${cleanupFunctionName}-lambda-role`, {
      roleName: `${cleanupFunctionName}-role`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });    
    lambdaRoleCleanup.attachInlinePolicy(new iam.Policy(this, `${cleanupFunctionName}-lambda-policy`, {
      policyName: `${cleanupFunctionName}-policy`,
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
          ],
          resources: [
            `arn:aws:logs:${cdk.Stack.of(this).region}:${accountId}:log-group:/aws/lambda/${props.pjName}-${props.envName}-*`,
          ],
        }),
      ],
    })); 
    lambdaRoleCleanup.attachInlinePolicy(new iam.Policy(this, `${cleanupFunctionName}-s3-policy`, {
      policyName: `${cleanupFunctionName}-s3-policy`,
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            's3:DeleteObject',
            's3:DeleteBucket',
          ],
          resources: [
            bucket.bucketArn,
            `${bucket.bucketArn}/*`,
          ],
        }),
      ],
    }));
    const cleanupFunctionPath = path.join(__dirname, '../src/lambda/cleanup/python');
    const cleanupFunction = new lambda.Function(this, cleanupFunctionName, {
      functionName: cleanupFunctionName,
      description: 'cleanup lambda function for s3',
      runtime: lambda.Runtime.PYTHON_3_13,
      handler: 'index.lambda_handler',
      timeout: cdk.Duration.minutes(15),
      memorySize: 1024,
      role: lambdaRoleCleanup,
      architecture: lambda.Architecture.ARM_64,
      code: lambda.Code.fromAsset(cleanupFunctionPath),
      environment: {
        BUCKET_NAME: bucket.bucketName,
      },
    });
    pipeline.addStage({
      stageName: 'Cleanup',
      actions: [
        new codepipeline_actions.LambdaInvokeAction({
          actionName: 'Cleanup',
          lambda: cleanupFunction,
          userParameters: {
            BUCKET_NAME: bucket.bucketName,
          },
        }),
      ],
    });

    // CloudFront Cache Invalidation
    if (!props.distributionId) {
      const cloudFrontInvalidationFunctionName = `${props.pjName}-${props.envName}-cloudfront-invalidation-lambda-${accountId}`;
      const cloudFrontInvalidationFunction = new lambda.Function(this, cloudFrontInvalidationFunctionName, {
        functionName: cloudFrontInvalidationFunctionName,
        description: 'CloudFront Cache Invalidation Lambda',
        runtime: lambda.Runtime.PYTHON_3_13,
        handler: 'index.lambda_handler',
        timeout: cdk.Duration.seconds(300),
        memorySize: 1024,
        role: lambdaRole,
        code: lambda.Code.fromAsset(path.join(__dirname, '../src/lambda/cloudfront-invalidation/python')),
        environment: {
          DISTRIBUTION_ID: '',
        },
      });
      cloudFrontInvalidationFunction.addPermission('CloudFrontInvalidationPermission', {
        principal: new iam.ServicePrincipal('lambda.amazonaws.com'),
        sourceArn: websiteBucket.bucketArn,
      });
      pipeline.addStage({
        stageName: 'CloudFrontInvalidation',
        actions: [
          new codepipeline_actions.LambdaInvokeAction({
            actionName: 'CloudFrontInvalidation',
            lambda: cloudFrontInvalidationFunction,
            userParameters: {
              DISTRIBUTION_ID: websiteBucket.bucketArn,
            },
          }),
        ],
      });
      cloudFrontInvalidationFunction.addEnvironment('DISTRIBUTION_ID', props.distributionId!);
    }
    // Add permissions for lambda to trigger pipeline
    const pipelineExecutionPolicy = new iam.Policy(this, 'PipelineExecutionPolicy', {
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'codepipeline:StartPipelineExecution',
          ],
          resources: [
            pipeline.pipelineArn,
          ],
        }),
      ],
    });
    lambdaRole.attachInlinePolicy(pipelineExecutionPolicy);

    // Update Lambda environment variables to include pipeline information
    gitCloneFunction.addEnvironment('PIPELINE_NAME', pipeline.pipelineName);
    gitCloneFunction.addEnvironment('ZIP_FILE_NAME', 'source.zip');

  }
}
