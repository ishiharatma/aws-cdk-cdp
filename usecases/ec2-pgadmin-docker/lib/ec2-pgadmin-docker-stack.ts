import * as cdk from 'aws-cdk-lib';
import { StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { 
  aws_ec2 as ec2,
  aws_events as events,
  aws_iam as iam,
} from 'aws-cdk-lib';

interface Ec2PgadminDockerStackProps extends StackProps {
  readonly pjName: string;
  readonly envName: string;
  readonly vpcId: string;
  readonly pgAdminLoginId: string;
  readonly pgAdminLoginPassword: string;
  readonly startSchedule?: string;
  readonly stopSchedule?: string;
  readonly isAutoDeleteObject?: boolean;
}

export class Ec2PgadminDockerStack extends cdk.Stack {
  public readonly instance: ec2.IInstance;
  constructor(scope: Construct, id: string, props: Ec2PgadminDockerStackProps) {
    super(scope, id, props);
    const accountId:string = cdk.Stack.of(this).account;
    const region:string = cdk.Stack.of(this).region;

    const vpc = ec2.Vpc.fromLookup(this, 'Vpc', {vpcId:props.vpcId});
    // Create a security group for EC2 instance
    const ec2SecurityGroup = new ec2.SecurityGroup(this, 'EC2SecurityGroup', {
      vpc: vpc,
      allowAllOutbound: true,    
    });
    new cdk.CfnOutput(this, 'EC2SecurityGroupId', {
      value: ec2SecurityGroup.securityGroupId,
    });
    
    const keyPair = new ec2.KeyPair(this, 'KeyPair', {
      keyPairName: `${id}-ec2-KeyPair`,
    });
    new cdk.CfnOutput(this, 'EC2InstanceKeyPairId', {
      value: keyPair.keyPairId,
    });
    // キーペア取得コマンドアウトプット
    new cdk.CfnOutput(this, 'GetSSHKeyCommand', {
      value: `aws ssm get-parameter --name /ec2/keypair/${keyPair.keyPairId} --region ${this.region} --with-decryption --query Parameter.Value --output text`,
    })
    const userData = ec2.UserData.forLinux({ shebang: '#!/bin/bash' });
    userData.addCommands(
      'yum update -y',
      'yum -y install docker',  // Docker Engine パッケージインストール
      'systemctl start docker', // Docker サービス起動
      'systemctl enable docker', // Dockerサービスの自動起動を有効化
      // PgAdmin4 のコンテナ起動
      `docker run --restart=always --name pgadmin4 -e "PGADMIN_DEFAULT_EMAIL=${props.pgAdminLoginId}" -e "PGADMIN_DEFAULT_PASSWORD=${props.pgAdminLoginPassword}" -d -p 5050:80 dpage/pgadmin4`,
    );

    // Create an EC2 instance for connection testing
    this.instance = new ec2.Instance(this, 'EC2Instance1', {
      vpc: vpc,
      instanceName: [props.pjName, props.envName, 'pgadmin', 'instance'].join('/') ,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T4G, ec2.InstanceSize.NANO),
      machineImage: ec2.MachineImage.latestAmazonLinux2023({
        edition: ec2.AmazonLinuxEdition.STANDARD,
        cpuType: ec2.AmazonLinuxCpuType.ARM_64 //X86_64, 
      }),
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
      securityGroup: ec2SecurityGroup,
      //availabilityZone: 'ap-northeast-1a',
      ssmSessionPermissions: true, // Used by SSM session manager
      userData: userData,
      // Configure IMDSv2 for the NAT instance
      // Security Hub EC2.8
      // https://docs.aws.amazon.com/ja_jp/securityhub/latest/userguide/ec2-controls.html#ec2-8
      requireImdsv2: true,
      keyPair: keyPair,
    });
    new cdk.CfnOutput(this, "EC2InstanceId", {
        value: this.instance.instanceId,
        description: "PgAdmin Instance",
    });

    // 自動起動と停止スケジュール
    if (props.startSchedule && props.stopSchedule) {
      const role = new iam.Role(this, `EC2startstopRole`, {
        roleName: [props.pjName, props.envName, 'PgAdminStopStart'].join('-'),
        assumedBy: new iam.ServicePrincipal('events.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonSSMAutomationRole'),
        ]
      });
      // 起動スケジュール
      const startRule = new events.CfnRule(this, 'EC2StartRule', {
        name: [props.pjName, props.envName, 'EC2StartRule'].join('-'),
        description: `${this.instance.instanceId} ${props.startSchedule} Start`,
        scheduleExpression: props.startSchedule,
        targets: [{
          arn: `arn:aws:ssm:${region}::automation-definition/AWS-StartEC2Instance:$DEFAULT`,
          id: 'TargetEC2Instance1',
          input: `{"InstanceId": ["${this.instance.instanceId}"]}`,
          roleArn: role.roleArn
        }]
      });
      
      // 停止スケジュール
      const stopRule = new events.CfnRule(this, 'EC2StopRule', {
        name: [props.pjName, props.envName, 'EC2StopRule'].join('-'),
        description: `${this.instance.instanceId} ${props.stopSchedule} Stop`,
        scheduleExpression: props.stopSchedule,
        targets: [{
          arn: `arn:aws:ssm:${region}::automation-definition/AWS-StopEC2Instance:$DEFAULT`,
          id: 'TargetEC2Instance1',
          input: `{"InstanceId": ["${this.instance.instanceId}"]}`,
          roleArn: role.roleArn
        }]
      });
    }

  }
}
