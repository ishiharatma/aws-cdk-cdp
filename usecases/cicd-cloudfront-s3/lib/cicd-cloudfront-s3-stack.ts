import * as cdk from 'aws-cdk-lib';
import { StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { WebPipelineConstruct } from '../../common/lib/construct-pipeline-s3web'

import * as interfaceDef from '../../common/interface/index';

interface CicdCloudFrontS3StackProps extends StackProps {
  input: interfaceDef.WebPipelineConstructProps;
}

export class CicdCloudfrontS3Stack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CicdCloudFrontS3StackProps) {
    super(scope, id, props);

    new WebPipelineConstruct(this, 'Default', props.input);

  }
}
