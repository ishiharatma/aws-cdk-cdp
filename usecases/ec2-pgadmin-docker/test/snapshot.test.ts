import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { App } from 'aws-cdk-lib';

import { Ec2PgadminDockerStack } from '../lib/ec2-pgadmin-docker-stack';

const defaultEnv = {
    account: '123456789012',
    region: 'ap-northeast-1',
};
const pjName: string = 'snapshot';
const envName: string = 'test';
const app = new App();

test('snapshot validation test',() =>{
    const stack = new Ec2PgadminDockerStack(app, 'VPCStack', {
        stackName:  `Ec2PgadminDockerStack-${pjName}-${envName}`,
        description: 'Setup EC2 for PgAdmin',
        pjName: pjName,
        envName: envName,
        vpcId: 'vpc-01234567890abcdef',
        pgAdminLoginId: 'pgadmin4@example.com',
        pgAdminLoginPassword : 'pasword',
        startSchedule: "cron(0 0 ? * MON-FRI *)" ,
        stopSchedule: "cron(0 9 ? * MON-FRI *)",
        isAutoDeleteObject: true,
        env: defaultEnv,
        terminationProtection: true, // Enabling deletion protection
      });
    // add tag
    cdk.Tags.of(app).add('Project', pjName);
    cdk.Tags.of(app).add('Environment', envName);
    // test with snapshot
    expect(Template.fromStack(stack)).toMatchSnapshot();

})