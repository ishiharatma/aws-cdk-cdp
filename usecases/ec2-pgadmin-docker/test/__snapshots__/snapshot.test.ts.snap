// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot validation test 1`] = `
{
  "Description": "Setup EC2 for PgAdmin",
  "Outputs": {
    "EC2InstanceId": {
      "Description": "PgAdmin Instance",
      "Value": {
        "Ref": "EC2Instance14196AB1A",
      },
    },
    "EC2InstanceKeyPairId": {
      "Value": {
        "Fn::GetAtt": [
          "KeyPair1622897B",
          "KeyPairId",
        ],
      },
    },
    "EC2SecurityGroupId": {
      "Value": {
        "Fn::GetAtt": [
          "EC2SecurityGroup05DEE054",
          "GroupId",
        ],
      },
    },
    "GetSSHKeyCommand": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "aws ssm get-parameter --name /ec2/keypair/",
            {
              "Fn::GetAtt": [
                "KeyPair1622897B",
                "KeyPairId",
              ],
            },
            " --region ap-northeast-1 --with-decryption --query Parameter.Value --output text",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestal2023amikernel61arm64C96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-6.1-arm64",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "EC2Instance14196AB1A": {
      "DependsOn": [
        "EC2Instance1InstanceRole34EB269B",
      ],
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "IamInstanceProfile": {
          "Ref": "EC2Instance1InstanceProfileF359155E",
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestal2023amikernel61arm64C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t4g.nano",
        "KeyName": {
          "Ref": "KeyPair1622897B",
        },
        "LaunchTemplate": {
          "LaunchTemplateName": "EC2Instance1LaunchTemplate",
          "Version": {
            "Fn::GetAtt": [
              "EC2Instance1LaunchTemplate1B7F9B1D",
              "LatestVersionNumber",
            ],
          },
        },
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "EC2SecurityGroup05DEE054",
              "GroupId",
            ],
          },
        ],
        "SubnetId": "p-12345",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "test",
          },
          {
            "Key": "Name",
            "Value": "snapshot/test/pgadmin/instance",
          },
          {
            "Key": "Project",
            "Value": "snapshot",
          },
        ],
        "UserData": {
          "Fn::Base64": "#!/bin/bash
yum update -y
yum -y install docker
systemctl start docker
systemctl enable docker
docker run --restart=always --name pgadmin4 -e "PGADMIN_DEFAULT_EMAIL=pgadmin4@example.com" -e "PGADMIN_DEFAULT_PASSWORD=pasword" -d -p 5050:80 dpage/pgadmin4",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "EC2Instance1InstanceProfileF359155E": {
      "Properties": {
        "Roles": [
          {
            "Ref": "EC2Instance1InstanceRole34EB269B",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "EC2Instance1InstanceRole34EB269B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Environment",
            "Value": "test",
          },
          {
            "Key": "Name",
            "Value": "snapshot/test/pgadmin/instance",
          },
          {
            "Key": "Project",
            "Value": "snapshot",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EC2Instance1LaunchTemplate1B7F9B1D": {
      "Properties": {
        "LaunchTemplateData": {
          "MetadataOptions": {
            "HttpTokens": "required",
          },
        },
        "LaunchTemplateName": "EC2Instance1LaunchTemplate",
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "EC2SecurityGroup05DEE054": {
      "Properties": {
        "GroupDescription": "VPCStack/EC2SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": [
          {
            "Key": "Environment",
            "Value": "test",
          },
          {
            "Key": "Project",
            "Value": "snapshot",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "EC2StartRule": {
      "Properties": {
        "Description": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "EC2Instance14196AB1A",
              },
              " cron(0 0 ? * MON-FRI *) Start",
            ],
          ],
        },
        "Name": "snapshot-test-EC2StartRule",
        "ScheduleExpression": "cron(0 0 ? * MON-FRI *)",
        "Targets": [
          {
            "Arn": "arn:aws:ssm:ap-northeast-1::automation-definition/AWS-StartEC2Instance:$DEFAULT",
            "Id": "TargetEC2Instance1",
            "Input": {
              "Fn::Join": [
                "",
                [
                  "{"InstanceId": ["",
                  {
                    "Ref": "EC2Instance14196AB1A",
                  },
                  ""]}",
                ],
              ],
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "EC2startstopRoleA8970F74",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "EC2StopRule": {
      "Properties": {
        "Description": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "EC2Instance14196AB1A",
              },
              " cron(0 9 ? * MON-FRI *) Stop",
            ],
          ],
        },
        "Name": "snapshot-test-EC2StopRule",
        "ScheduleExpression": "cron(0 9 ? * MON-FRI *)",
        "Targets": [
          {
            "Arn": "arn:aws:ssm:ap-northeast-1::automation-definition/AWS-StopEC2Instance:$DEFAULT",
            "Id": "TargetEC2Instance1",
            "Input": {
              "Fn::Join": [
                "",
                [
                  "{"InstanceId": ["",
                  {
                    "Ref": "EC2Instance14196AB1A",
                  },
                  ""]}",
                ],
              ],
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "EC2startstopRoleA8970F74",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "EC2startstopRoleA8970F74": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonSSMAutomationRole",
              ],
            ],
          },
        ],
        "RoleName": "snapshot-test-PgAdminStopStart",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "test",
          },
          {
            "Key": "Project",
            "Value": "snapshot",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "KeyPair1622897B": {
      "Properties": {
        "KeyFormat": "pem",
        "KeyName": "VPCStack-ec2-KeyPair",
        "KeyType": "rsa",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "test",
          },
          {
            "Key": "Project",
            "Value": "snapshot",
          },
        ],
      },
      "Type": "AWS::EC2::KeyPair",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
